provider/waiter.go

```diff
@@ -188,13 +188,13 @@ func (w *NoopWaiter) Wait(_ context.Context) error {
 // FieldPathToTftypesPath takes a string representation of
 // a path to a field in dot/square bracket notation
 // and returns a tftypes.AttributePath
-func FieldPathToTftypesPath(fieldPath string) (tftypes.AttributePath, error) {
+func FieldPathToTftypesPath(fieldPath string) (*tftypes.AttributePath, error) {
 	t, d := hclsyntax.ParseTraversalAbs([]byte(fieldPath), "", hcl.Pos{Line: 1, Column: 1})
 	if d.HasErrors() {
-		return tftypes.AttributePath{}, fmt.Errorf("invalid field path %q: %s: %s", fieldPath, d[0].Summary, d[0].Detail)
+		return tftypes.NewAttributePath(), fmt.Errorf("invalid field path %q: %s: %s", fieldPath, d[0].Summary, d[0].Detail)
```

Comment:
My IDE flagged this line as a potential source of a panic, if d[0] is nil. I tested it out in a Go REPL and found I could trigger a panic if I passed a bad value in for `fieldPath`. Here's the full output of my test, in case you're interested:

```
gore> :import "github.com/hashicorp/hcl/v2/hclsyntax"
gore> :import "github.com/hashicorp/hcl/v2"
gore> t, d := hclsyntax.ParseTraversalAbs([]byte("potato"), "", hcl.Pos{Line: 1, Column: 1})

hcl.Traversal{
  hcl.TraverseRoot{
    isTraverser: hcl.isTraverser{},
    Name:        "potato",
    SrcRange:    hcl.Range{
      Filename: "",
      Start:    hcl.Pos{
        Line:   1,
        Column: 1,
        Byte:   0,
      },
      End: hcl.Pos{
        Line:   1,
        Column: 7,
        Byte:   6,
      },
    },
  },
}
hcl.Diagnostics(nil)

gore> d[0].Detail
panic: runtime error: index out of range [0] with length 0
```

Reply:
Thanks for spotting this, @dak1n1 !

Indeed, this has potential to blow up in more than one way. The type of `d` as returned by `ParseTraversalAbs` is `Diagnostics` which is defined as slice of pointers to Diagnostic. So the panic can either come from indexing an empty slice, as you've illustrated in your example, or in case the first element in the slice is a nil pointer to Diagnostic (in which case `d[0].Summary` will panic).

The empty slice situation is guarded against by this line of code `if d.HasErrors() {` which will ensure the slice is not empty. However, it doesn't guarantee that the first element in Diagnostics is the actual error. So I changed that line to use the `Errors()` method from `Diagnostics` itself, which I think is how they intended this to be used.

Finally, the other panic mode is when dereferencing a nil pointer (e.g. `d[0] == nil`). Naturally, we could check it before using it, but having a quick peek inside `hclsyntax.ParseTraversalAbs()` it looks like it will never produce nil elements in Diagnostics. The fact that its own `Errors()` method also uses the diagnostic elements without nil checks confirms that.